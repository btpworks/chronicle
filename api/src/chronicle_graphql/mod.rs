use async_graphql::{
    extensions::OpenTelemetry,
    http::{playground_source, GraphQLPlaygroundConfig},
    Context, Enum, Error, ErrorExtensions, Object, ObjectType, Schema, SimpleObject, Subscription,
};
use async_graphql_poem::{GraphQL, GraphQLSubscription};
use chrono::{DateTime, NaiveDateTime, Utc};
use custom_error::custom_error;
use derivative::*;
use diesel::{
    prelude::*,
    r2d2::{ConnectionManager, Pool},
    Queryable, SqliteConnection,
};
use futures::Stream;
use poem::{
    get, handler, listener::TcpListener, post, web::Html, EndpointExt, IntoResponse, Route, Server,
};
use std::{net::SocketAddr, time::Duration};
use tokio::sync::broadcast::error::RecvError;

use crate::ApiDispatch;
#[macro_use]
mod cursor_query;
pub mod activity;
pub mod agent;
pub mod entity;
pub mod mutation;
pub mod query;

#[derive(Default, Queryable, Selectable, SimpleObject)]
#[diesel(table_name = crate::persistence::schema::agent)]
pub struct Agent {
    pub id: i32,
    pub name: String,
    pub namespace_id: i32,
    pub domaintype: Option<String>,
    pub current: i32,
    pub identity_id: Option<i32>,
}

#[derive(Default, Queryable, Selectable)]
#[diesel(table_name = crate::persistence::schema::identity)]
pub struct Identity {
    pub id: i32,
    pub namespace_id: i32,
    pub public_key: String,
}

#[Object]
impl Identity {
    async fn public_key(&self) -> &str {
        &self.public_key
    }
}

#[derive(Default, Queryable, Selectable, SimpleObject)]
#[diesel(table_name = crate::persistence::schema::activity)]
pub struct Activity {
    pub id: i32,
    pub name: String,
    pub namespace_id: i32,
    pub domaintype: Option<String>,
    pub started: Option<NaiveDateTime>,
    pub ended: Option<NaiveDateTime>,
}

#[derive(Queryable, Selectable, SimpleObject)]
#[diesel(table_name = crate::persistence::schema::entity)]
pub struct Entity {
    pub id: i32,
    pub name: String,
    pub namespace_id: i32,
    pub domaintype: Option<String>,
    pub attachment_id: Option<i32>,
}

#[derive(Queryable, Selectable)]
#[diesel(table_name = crate::persistence::schema::attachment)]
#[allow(dead_code)]
pub struct Evidence {
    id: i32,
    namespace_id: i32,
    signature_time: NaiveDateTime,
    signature: String,
    signer_id: i32,
    locator: Option<String>,
}

#[Object]
impl Evidence {
    async fn signature_time(&self) -> DateTime<Utc> {
        DateTime::from_utc(self.signature_time, Utc)
    }

    async fn signature(&self) -> &str {
        &self.signature
    }

    async fn locator(&self) -> Option<&str> {
        self.locator.as_deref()
    }
}

#[derive(Default, Queryable)]
pub struct Namespace {
    _id: i32,
    uuid: String,
    name: String,
}

#[Object]
impl Namespace {
    async fn name(&self) -> &str {
        &self.name
    }

    async fn uuid(&self) -> &str {
        &self.uuid
    }
}

#[derive(Default, Queryable)]
pub struct Submission {
    context: String,
    correlation_id: String,
}

#[Object]
impl Submission {
    async fn context(&self) -> &str {
        &self.context
    }

    async fn correlation_id(&self) -> &str {
        &self.correlation_id
    }
}

#[derive(Enum, Copy, Clone, Eq, PartialEq, Debug)]
pub enum TimelineOrder {
    NewestFirst,
    OldestFirst,
}

custom_error! {pub GraphQlError
    Db{source: diesel::result::Error}                           = "Database operation failed",
    R2d2{source: r2d2::Error }                                  = "Connection pool error",
    DbConnection{source: diesel::ConnectionError}               = "Database connection failed",
    Api{source: crate::ApiError}                                = "API",
    Io{source: std::io::Error}                                  = "I/O",
}

impl GraphQlError {
    fn error_sources(
        mut source: Option<&(dyn std::error::Error + 'static)>,
    ) -> Option<Vec<String>> {
        /* Check if we have any sources to derive reasons from */
        if source.is_some() {
            /* Add all the error sources to a list of reasons for the error */
            let mut reasons = Vec::new();
            while let Some(error) = source {
                reasons.push(error.to_string());
                source = error.source();
            }
            Some(reasons)
        } else {
            None
        }
    }
}

impl ErrorExtensions for GraphQlError {
    // lets define our base extensions
    fn extend(&self) -> Error {
        Error::new(self.to_string()).extend_with(|_err, e| {
            if let Some(reasons) = Self::error_sources(custom_error::Error::source(&self)) {
                let mut i = 1;
                for reason in reasons {
                    e.set(format!("reason {}", i), reason);
                    i += 1;
                }
            }
        })
    }
}

#[derive(Derivative)]
#[derivative(Debug)]
pub struct Store {
    #[derivative(Debug = "ignore")]
    pub pool: Pool<ConnectionManager<SqliteConnection>>,
}

impl Store {
    pub fn new(pool: Pool<ConnectionManager<SqliteConnection>>) -> Self {
        Store { pool }
    }
}

pub struct Subscription;

#[derive(Queryable)]
pub struct CommitNotification {
    correlation_id: String,
}

#[Object]
impl CommitNotification {
    pub async fn correlation_id(&self) -> &String {
        &self.correlation_id
    }
}

#[Subscription]
impl Subscription {
    async fn commit_notifications<'a>(
        &self,
        ctx: &Context<'a>,
    ) -> impl Stream<Item = CommitNotification> {
        let api = ctx.data_unchecked::<ApiDispatch>().clone();
        let mut rx = api.notify_commit.subscribe();
        async_stream::stream! {
            loop {
                match rx.recv().await {
                    Ok((_prov, correlation_id)) =>
                    yield CommitNotification {correlation_id: correlation_id.to_string()},
                    Err(RecvError::Lagged(_)) => {
                    }
                    Err(_) => break
                }
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct ChronicleGraphQl<Query, Mutation>
where
    Query: ObjectType + 'static,
    Mutation: ObjectType + 'static,
{
    query: Query,
    mutation: Mutation,
}

#[async_trait::async_trait]
pub trait ChronicleGraphQlServer {
    async fn serve_graphql(
        &self,
        pool: Pool<ConnectionManager<SqliteConnection>>,
        api: ApiDispatch,
        address: SocketAddr,
        open: bool,
    );
}

impl<Query, Mutation> ChronicleGraphQl<Query, Mutation>
where
    Query: ObjectType + Copy,
    Mutation: ObjectType + Copy,
{
    pub fn new(query: Query, mutation: Mutation) -> Self {
        Self { query, mutation }
    }

    pub fn exportable_schema(&self) -> String
    where
        Query: ObjectType + Copy,
        Mutation: ObjectType + Copy,
    {
        let schema = Schema::build(self.query, self.mutation, Subscription).finish();

        schema.sdl()
    }
}

#[handler]
async fn gql_playground() -> impl IntoResponse {
    Html(playground_source(
        GraphQLPlaygroundConfig::new("/").subscription_endpoint("/ws"),
    ))
}

#[async_trait::async_trait]
impl<Query, Mutation> ChronicleGraphQlServer for ChronicleGraphQl<Query, Mutation>
where
    Query: ObjectType + Copy,
    Mutation: ObjectType + Copy,
{
    async fn serve_graphql(
        &self,
        pool: Pool<ConnectionManager<SqliteConnection>>,
        api: ApiDispatch,
        address: SocketAddr,
        open: bool,
    ) {
        let schema = Schema::build(self.query, self.mutation, Subscription)
            .extension(OpenTelemetry::new(opentelemetry::global::tracer(
                "chronicle-api-gql",
            )))
            .data(Store::new(pool.clone()))
            .data(api)
            .finish();

        if open {
            tokio::spawn(async move {
                tokio::time::sleep(Duration::from_millis(200)).await;
                open::that(format!("http://{}", address)).ok();
            });
            let app = Route::new()
                .at("/", get(gql_playground).post(GraphQL::new(schema.clone())))
                .at("/ws", get(GraphQLSubscription::new(schema.clone())))
                .data(schema.clone());

            Server::new(TcpListener::bind(address)).run(app).await.ok();
        } else {
            let app = Route::new()
                .at("/", post(GraphQL::new(schema.clone())))
                .at("/ws", get(GraphQLSubscription::new(schema)));

            Server::new(TcpListener::bind(address)).run(app).await.ok();
        }
    }
}
