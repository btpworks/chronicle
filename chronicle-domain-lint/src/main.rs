use chronicle::codegen::model;
use jsonschema::JSONSchema;
use std::{collections::HashSet, env, path::Path, process::exit};

fn bad_filename(filename: &str) {
    println!("JSON or YAML filename extension required for {}", filename);
    exit(2);
}

fn build_json_validator(domain: &str) -> JSONSchema {
    let json = match serde_json::from_str(domain) {
        Ok(json) => json,
        Err(error) => {
            println!("failed to parse valid JSON from domain schema: {}", error);
            exit(2);
        }
    };
    match JSONSchema::options()
        .with_draft(jsonschema::Draft::Draft7)
        .compile(&json)
    {
        Ok(json_schema) => json_schema,
        Err(error) => {
            println!("failed to interpret JSON as a domain schema: {}", error);
            exit(2);
        }
    }
}

fn check_json_valid(json_validator: &JSONSchema, json_data: &str) {
    let json = match serde_json::from_str(json_data) {
        Ok(json) => json,
        Err(error) => {
            println!("failed to parse valid JSON: {}", error);
            exit(2);
        }
    };
    let validation = json_validator.validate(&json);
    if let Err(mut errors) = validation {
        match errors.next() {
            Some(error) => println!(
                "path {} contains invalid data: {}",
                error.instance_path, error
            ),
            None => println!("failed to validate against schema"),
        };
        exit(2);
    }
}

fn check_yaml_valid(json_validator: &JSONSchema, yaml_data: &str) {
    let json = match serde_yaml::from_str::<serde_json::Value>(yaml_data) {
        Ok(json) => json,
        Err(error) => {
            println!("failed to parse valid YAML: {}", error);
            exit(2);
        }
    };
    let json_data = match serde_json::to_string(&json) {
        Ok(json_data) => json_data,
        Err(error) => {
            println!("failed to write valid JSON from YAML: {}", error);
            exit(2);
        }
    };
    check_json_valid(json_validator, &json_data);
}

fn read_json_domain(data: &str) -> model::DomainFileInput {
    match serde_json::from_str(data) {
        Ok(domain) => domain,
        Err(error) => {
            println!("failed to interpret JSON as a domain: {}", error);
            exit(2);
        }
    }
}

fn read_yaml_domain(data: &str) -> model::DomainFileInput {
    match serde_yaml::from_str(data) {
        Ok(domain) => domain,
        Err(error) => {
            println!("failed to interpret YAML as a domain: {}", error);
            exit(2);
        }
    }
}

fn check_domain_attributes(
    element: &str,
    attributes: &HashSet<String>,
    named_resources: Vec<(&String, &model::ResourceDef)>,
) {
    for (name, resource) in named_resources {
        for attribute in resource.attributes.iter() {
            if !(attributes.contains(&attribute.0)) {
                println!(
                    "{} named {} has unknown attribute {}",
                    element, name, attribute.0
                );
                exit(2);
            }
        }
    }
}

fn check_domain(domain: model::DomainFileInput) {
    let attributes = domain
        .attributes
        .keys()
        .map(std::clone::Clone::clone)
        .collect();
    check_domain_attributes("agent", &attributes, domain.agents.iter().collect());
    check_domain_attributes("entity", &attributes, domain.entities.iter().collect());
    check_domain_attributes("activity", &attributes, domain.activities.iter().collect());
}

fn main() {
    let filenames = env::args().collect::<Vec<String>>();
    if filenames.len() < 2 {
        println!("command argument(s) should be filename(s) of domain definitions");
        exit(1);
    }
    let json_validator = build_json_validator(include_str!("../schema/domain.json"));
    for filename in filenames.get(1..).unwrap() {
        let filepath = Path::new(filename);
        let data = match std::fs::read_to_string(filepath) {
            Ok(data) => data,
            Err(error) => {
                println!("failed to read {}: {}", filename, error);
                exit(2);
            }
        };
        match filepath.extension() {
            Some(extension) => {
                match extension.to_ascii_lowercase().to_str() {
                    Some("json") | Some("jsn") => {
                        check_json_valid(&json_validator, data.as_str());
                        check_domain(read_json_domain(&data));
                    }
                    Some("yaml") | Some("yml") => {
                        check_yaml_valid(&json_validator, data.as_str());
                        check_domain(read_yaml_domain(&data));
                    }
                    _ => {
                        bad_filename(filename);
                    }
                };
            }
            None => {
                bad_filename(filename);
            }
        };
        println!(
            "successful: {} appears to contain a valid domain definition",
            filename
        );
    }
}
