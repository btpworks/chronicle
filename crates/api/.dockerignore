sqlite_test
db.sqlite
