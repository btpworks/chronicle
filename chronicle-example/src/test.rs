use chronicle::codegen::ChronicleDomainDef;
use chronicle::tokio;
use chronicle::{api::chronicle_graphql::ChronicleGraphQl, bootstrap};
use main::{Mutation, Query};

#[allow(dead_code)]
mod main;

///Entry point here is jigged a little, as we want to run unit tests, see chronicle-untyped for the actual pattern
#[tokio::main]
pub async fn main() {
    let s = r#"
name: "chronicle"
attributes:
  Title:
    type: "String"
  Location:
    type: "String"
  PurchaseValue:
    type: "String"
  PurchaseValueCurrency:
    type: "String"
  Description:
    type: "String"
  Name:
    type: "String"
agents:
  Member:
    attributes:
      - Name
  Artist:
    attributes:
      - Name
entities:
  Artwork:
    attributes:
      - Title
  ArtworkDetails:
    attributes:
      - Title
      - Description
activities:
  Exhibited:
    attributes:
      - Location
  Created:
    attributes:
      - Title
  Sold:
    attributes:
      - PurchaseValue
      - PurchaseValueCurrency
  Transferred:
    attributes: []
roles:
  - Buyer
  - Seller
  - Broker
  - Editor
  - Creator
"#
    .to_string();

    let model = ChronicleDomainDef::from_input_string(&s).unwrap();

    bootstrap(model, ChronicleGraphQl::new(Query, Mutation)).await
}

#[cfg(test)]
mod test {
    use super::{Mutation, Query};
    use chronicle::api::chronicle_graphql::{Store, Subscription};
    use chronicle::api::{Api, ConnectionOptions, UuidGen};
    use chronicle::async_graphql::{Request, Schema};
    use chronicle::chrono::{DateTime, NaiveDate, Utc};
    use chronicle::common::ledger::InMemLedger;
    use chronicle::tokio;
    use chronicle::uuid::Uuid;
    use diesel::r2d2::Pool;
    use diesel::{r2d2::ConnectionManager, SqliteConnection};
    use std::time::Duration;
    use tempfile::TempDir;

    #[derive(Debug, Clone)]
    struct SameUuid;

    impl UuidGen for SameUuid {
        fn uuid() -> Uuid {
            Uuid::parse_str("5a0ab5b8-eeb7-4812-9fe3-6dd69bd20cea").unwrap()
        }
    }

    async fn test_schema() -> Schema<Query, Mutation, Subscription> {
        telemetry::telemetry(None, true);

        let secretpath = TempDir::new().unwrap();

        // We need to use a real file for sqlite, as in mem either re-creates between
        // macos temp dir permissions don't work with sqlite
        std::fs::create_dir("./sqlite_test").ok();
        let dbid = Uuid::new_v4();
        let mut ledger = InMemLedger::new();
        let reader = ledger.reader();

        let pool = Pool::builder()
            .connection_customizer(Box::new(ConnectionOptions {
                enable_wal: true,
                enable_foreign_keys: true,
                busy_timeout: Some(Duration::from_secs(2)),
            }))
            .build(ConnectionManager::<SqliteConnection>::new(&*format!(
                "./sqlite_test/db{}.sqlite",
                dbid
            )))
            .unwrap();

        let dispatch = Api::new(
            pool.clone(),
            ledger,
            reader,
            &secretpath.into_path(),
            SameUuid,
        )
        .await
        .unwrap();

        Schema::build(Query, Mutation, Subscription)
            .data(Store::new(pool))
            .data(dispatch)
            .finish()
    }

    #[tokio::test]
    async fn agent_delegation() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                actedOnBehalfOf(
                    responsible: "http://blockchaintp.com/chronicle/ns#agent:responsible",
                    delegate: "http://blockchaintp.com/chronicle/ns#agent:delegate",
                    role: DELEGATE
                    ) {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;

        let derived = schema
            .execute(Request::new(
                r#"
            query {
                agentById(id: "http://blockchaintp.com/chronicle/ns#agent:responsible") {
                    ... on ProvAgent {
                        id
                        name
                        actedOnBehalfOf {
                            agent {
                                ... on ProvAgent {
                                    id
                                }
                            }
                            role
                        }
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_json_snapshot!(derived.data);
    }

    #[tokio::test]
    async fn untyped_derivation() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                wasDerivedFrom(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                               usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {
                    ... on ProvEntity {
                        id
                        name
                        wasDerivedFrom {
                            ... on ProvEntity {
                                id
                            }
                        }
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn primary_source() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                hadPrimarySource(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                               usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {

                    ... on ProvEntity {
                        id
                        name
                        wasDerivedFrom {
                            ... on ProvEntity {
                                id
                            }
                        }
                        hadPrimarySource{
                            ... on ProvEntity {
                                id
                            }
                        }
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn revision() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                wasRevisionOf(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                            usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {
                    ... on ProvEntity {
                        id
                        name
                        wasDerivedFrom {
                            ... on ProvEntity {
                                id
                            }
                        }
                        wasRevisionOf {
                            ... on ProvEntity {
                                id
                            }
                        }
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn quotation() {
        let schema = test_schema().await;

        let created = schema
            .execute(Request::new(
                r#"
            mutation {
                wasQuotedFrom(generatedEntity: "http://blockchaintp.com/chronicle/ns#entity:generated",
                            usedEntity: "http://blockchaintp.com/chronicle/ns#entity:used") {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(created);

        tokio::time::sleep(Duration::from_millis(100)).await;
        let derived = schema
            .execute(Request::new(
                r#"
            query {
                entityById(id: "http://blockchaintp.com/chronicle/ns#entity:generated") {
                    ... on ProvEntity {
                        id
                        name
                        wasDerivedFrom {
                            ... on ProvEntity {
                                id
                            }
                        }
                        wasQuotedFrom {
                            ... on ProvEntity {
                                id
                            }
                        }
                    }
                }
            }
        "#,
            ))
            .await;
        insta::assert_toml_snapshot!(derived);
    }

    #[tokio::test]
    async fn agent_can_be_created() {
        let schema = test_schema().await;

        let create = schema
            .execute(Request::new(
                r#"
            mutation {
                agent(name:"bobross", attributes: { type: "artist" }) {
                    context
                }
            }
        "#,
            ))
            .await;

        insta::assert_toml_snapshot!(create);
    }

    #[tokio::test]
    async fn query_activity_timeline() {
        let schema = test_schema().await;

        let res = schema
                .execute(Request::new(
                    r#"
            mutation {
                friend(name:"ringo", attributes: { stringAttribute: "string", intAttribute: 1, boolAttribute: false }) {
                    context
                }
            }
        "#,
                ))
                .await;

        assert_eq!(res.errors, vec![]);

        tokio::time::sleep(Duration::from_millis(100)).await;

        let res = schema
                .execute(Request::new(
                    r#"
            mutation {
                theSea(name:"coral", attributes: { stringAttribute: "string", intAttribute: 1, boolAttribute: false }) {
                    context
                }
            }
        "#,
                ))
                .await;

        assert_eq!(res.errors, vec![]);

        tokio::time::sleep(Duration::from_millis(100)).await;

        let from = DateTime::<Utc>::from_utc(NaiveDate::from_ymd(1968, 9, 1).and_hms(0, 0, 0), Utc);

        for i in 1..10 {
            let res = schema
                .execute(Request::new(format!(
                    r#"
            mutation {{
                gardening(name:"gardening{}", attributes: {{ stringAttribute: "String", intAttribute: 1, boolAttribute: false }}) {{
                    context
                }}
            }}
        "#,
                    i
                )))
                .await;
            assert_eq!(res.errors, vec![]);

            tokio::time::sleep(Duration::from_millis(100)).await;
            let res = schema
                .execute(Request::new(format!(
                    r#"
            mutation {{
                used(id: "http://blockchaintp.com/chronicle/ns#entity:coral", activity: "http://blockchaintp.com/chronicle/ns#activity:gardening{}") {{
                    context
                }}
            }}
        "#,
                    i
                )))
                .await;
            assert_eq!(res.errors, vec![]);

            tokio::time::sleep(Duration::from_millis(100)).await;

            let res = schema
                .execute(Request::new(format!(
                    r#"
            mutation {{
                endActivity( time: "{}", id: "http://http://blockchaintp.com/chronicle/ns#activity:gardening{}") {{
                    context
                }}
            }}
        "#,
                   from.checked_add_signed(chronicle::chrono::Duration::days(i)).unwrap().to_rfc3339() ,i
                )))
                .await;

            assert_eq!(res.errors, vec![]);

            tokio::time::sleep(Duration::from_millis(100)).await;

            let res = schema
                .execute(Request::new(format!(
                    r#"
            mutation {{
                wasAssociatedWith( role: RESPONSIBLE, responsible: "http://blockchaintp.com/chronicle/ns#agent:ringo", activity: "http://http://blockchaintp.com/chronicle/ns#activity:gardening{}") {{
                    context
                }}
            }}
        "#, i
                )))
                .await;

            tokio::time::sleep(Duration::from_millis(100)).await;

            assert_eq!(res.errors, vec![]);
        }

        tokio::time::sleep(Duration::from_millis(100)).await;

        let default_cursor = schema
            .execute(Request::new(
                r#"
                query {
                activityTimeline(forEntity: ["http://blockchaintp.com/chronicle/ns#entity:coral"],
                                from: "1968-01-01T00:00:00Z",
                                to: "2030-01-01T00:00:00Z",
                                activityTypes: [GARDENING],
                                ) {
                    pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                    }
                    edges {
                        node {
                            __typename
                            ... on Gardening {
                                id
                                name
                                stringAttribute
                                intAttribute
                                boolAttribute
                                wasAssociatedWith {
                                        responsible {
                                            agent {
                                                ... on Friend {
                                                    id
                                                    name
                                                }
                                            }
                                            role
                                        }
                                }
                                used {
                                    ... on TheSea {
                                        id
                                        name
                                    }
                                }
                            }
                       }
                        cursor
                    }
                }
                }
        "#,
            ))
            .await;

        insta::assert_json_snapshot!(default_cursor);
    }

    #[tokio::test]
    async fn query_agents_by_cursor() {
        let schema = test_schema().await;

        for i in 0..100 {
            let res = schema
                .execute(Request::new(format!(
                    r#"
            mutation {{
                friend(name:"bobross{}", attributes: {{ stringAttribute: "String", intAttribute: 1, boolAttribute: false }}) {{
                    context
                }}
            }}
        "#,
                    i
                )))
                .await;

            assert_eq!(res.errors, vec![]);
        }
        tokio::time::sleep(Duration::from_millis(100)).await;

        let default_cursor = schema
            .execute(Request::new(
                r#"
                query {
                agentsByType(agentType: FRIEND) {
                    pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                    }
                    edges {
                        node {
                            __typename
                            ... on Friend {
                                id
                                name
                                stringAttribute
                                intAttribute
                                boolAttribute
                            }
                       }
                        cursor
                    }
                }
                }
        "#,
            ))
            .await;

        insta::assert_json_snapshot!(default_cursor);

        let middle_cursor = schema
            .execute(Request::new(
                r#"
                query {
                agentsByType(agentType: FRIEND, first: 20, after: "3") {
                    pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                    }
                    edges {
                        node {
                            __typename
                            ... on Friend {
                                id
                                name
                                stringAttribute
                                intAttribute
                                boolAttribute
                            }
                        }
                        cursor
                    }
                }
                }
        "#,
            ))
            .await;

        insta::assert_json_snapshot!(middle_cursor);

        let out_of_bound_cursor = schema
            .execute(Request::new(
                r#"
                query {
                agentsByType(agentType: FRIEND, first: 20, after: "90") {
                    pageInfo {
                        hasPreviousPage
                        hasNextPage
                        startCursor
                        endCursor
                    }
                    edges {
                        node {
                            __typename
                            ... on Friend {
                                id
                                name
                                stringAttribute
                                intAttribute
                                boolAttribute
                            }
                        }
                        cursor
                    }
                }
                }
        "#,
            ))
            .await;

        insta::assert_json_snapshot!(out_of_bound_cursor);
    }
}
