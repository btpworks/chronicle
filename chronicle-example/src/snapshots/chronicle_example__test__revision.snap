---
source: chronicle-example/src/test.rs
assertion_line: 220
expression: created
---
[data.wasRevisionOf]
context = 'http://blockchaintp.com/chronicle/ns#entity:generated'
